Workspaces

TinyMarsh opened the pull request:
# Description

This PR implements some workspaces stuff, hopefully addressing #4.

The code is a bit messy but hopefully the basic stuff is there. To test, fetch a list of available workspaces and their IDs by running
```sh
clockify-tui list-workspaces
```
After including the ID in the config file you can run
```sh
clockify-tui workspace-info
```
to fetch information about the workspace ID you provided in the config.

The workspace info thing might be a bit useless, but I suppose it just serves to demonstrate that it works. Having this ability to fetch and set workspace IDs will come in handy later when implementing further functionality.

Fixes #4 

## Type of change

- [ ] Documentation (non-breaking change that adds or improves the documentation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking, back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Key checklist

- [x] All tests pass (eg. `python -m pytest`)
- [ ] The documentation builds and looks OK (eg. `python -m sphinx -b html docs docs/build`)
- [ ] Pre-commit hooks run successfully (eg. `pre-commit run --all-files`)

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added or an issue has been opened to tackle that in the future. (Indicate issue here: # (issue))

TinyMarsh committed:
Adds simple client connection and fetching workspaces

TinyMarsh committed:
Ability to set workspace_id

alexdewar reviewed (changes requested):
This looks useful. Thanks!

One problem at the mo though is that users have to specify a non-empty `workspace_id` in the config file otherwise it'll fail validation, which means they can't run the `list-workspaces` command to get a list of workspace IDs unless they already have one in the config file -- a bit of a chicken/egg situation. How about if users don't specify a workspace ID, we just use the first one returned by `client.workspaces.get_workspaces()`?

A related problem is that because `try_read_config()` is called when initialising a couple of the modules, if you try to run any command without a config file present (e.g. you want to create one with `clockify-tui edit-config`) then it won't work. The way to get round this is just to call it where it's needed, e.g. in `workspace_info()`.

TinyMarsh committed:
Use first workspace ID if none provided

TinyMarsh commented:
I've addressed the first half of your comment, but I'm not really sure how we want to go about the `try_read_config()` problem. If no config is present then the user is prompted to create one, if the user selects no then it fails. The problem is we always need to have a config file because we always need at least the `API_KEY`. It seems that we shouldn't really be offering the user the option to not create a config file?

alexdewar committed:
Rename: connect.py => clockify.py

alexdewar committed:
Fix: Don't try to load the config file on every code path

Some commands, such as `edit-config` or `--help` need to be able to run without a config file present.

alexdewar commented:
Sorry, I thought I replied to this and now I've noticed you're away :-(.

The issue is that while *almost* every command needs a valid config file for the program to work, there are a couple of cases where we need the command to run even without a config file (or a malformed one). For example, we want the user to be able to run `clockify-tui --help` without one and that won't currently work.

As you're away, I'll just finish this up so we can make use of it today. Hope that's ok!

alexdewar committed:
Merge remote-tracking branch 'origin/main' into workspaces

alexdewar committed:
Fix: mypy: Set ignore_missing_imports
